### PR TITLE
fix(ci): harden formal dispatch inputs

### LINKED ISSUE #3378: Harden formal verification workflow_dispatch inputs before shell use
# Harden formal verification workflow_dispatch inputs before shell use

## Summary

The formal verification workflow accepts workflow_dispatch parameters such as `tlaFile` and `alloyJar`. The Apalache job interpolates `tlaFile` directly into a shell `run` command, and the Alloy runner builds a shell command template that includes `ALLOY_JAR` derived from the dispatch input.

## Affected area

- `.github/workflows/formal-verify.yml`
- `scripts/formal/verify-alloy.mjs`

## Security impact

If an untrusted or prompt-influenced automation path can set these dispatch inputs, the values can influence shell execution on the GitHub Actions runner. The exact impact depends on repository workflow-dispatch permissions, token defaults, selected ref, and secret availability, which were not queried during the local audit.

## Recommended remediation

- Use typed `choice` inputs for finite values such as target, engine, and solver.
- Constrain file and jar parameters to approved repository-relative paths, or remove them from workflow_dispatch.
- Do not embed `${{ inputs.* }}` directly in `run` shell blocks; pass validated values through environment variables to a checked-in script.
- Replace the Alloy shell command template with an argv-safe `java` invocation.
- Add explicit least-privilege workflow permissions and an approval/actor allowlist for high-impact manual formal runs.

## Validation performed

Static source-to-sink review only. No workflow dispatch, GitHub API call, shell payload execution, dependency installation, external scan, or network request was performed.

### PR DESCRIPTION
**Pull request by ootakazuhiko**
## Summary

- Harden `formal-verify.yml` workflow_dispatch inputs by converting finite values to `choice`, removing manual jar path dispatch inputs, and disabling persisted checkout credentials.
- Add a workflow_dispatch actor authorization script requiring write-like collaborator permissions before manual formal jobs run repository scripts.
- Add shared formal file input policy for approved repository-relative paths, and apply it to TLA, Apalache, and Alloy runners.
- Disable `ALLOY_RUN_CMD` shell-template execution in `scripts/formal/verify-alloy.mjs`; Alloy jar execution now uses argv-safe `java` arguments.
- Update formal run documentation and add regression tests for workflow dispatch boundaries and formal input validation.

Fixes #3378.

## Validation

- `pnpm -s exec vitest run tests/unit/ci/formal-verify-workflow-security.test.ts tests/unit/ci/assert-workflow-dispatch-actor.test.ts tests/unit/formal/input-policy.test.ts tests/unit/formal/verify-alloy-security.test.ts tests/unit/formal/verify-tla-input-security.test.ts --reporter dot`
- `pnpm -s exec eslint --quiet scripts/formal/input-policy.mjs scripts/formal/verify-alloy.mjs scripts/formal/verify-tla.mjs scripts/formal/verify-apalache.mjs scripts/ci/assert-workflow-dispatch-actor.mjs tests/unit/ci/formal-verify-workflow-security.test.ts tests/unit/ci/assert-workflow-dispatch-actor.test.ts tests/unit/formal/input-policy.test.ts tests/unit/formal/verify-alloy-security.test.ts tests/unit/formal/verify-tla-input-security.test.ts`
- `pnpm -s run check:doc-consistency`
- `pnpm -s run types:check`
- `pnpm -s run check:schemas`
- `git diff --check`

Local note: `pnpm -s run lint:actions` could not complete in this environment because the configured podman actionlint container failed with `cannot re-exec process to join the existing user namespace` (container exit 125). The workflow is covered by YAML parse/regression tests locally and will be validated by CI actionlint.

## Acceptance

- Manual formal dispatch finite inputs are typed choices.
- `tlaFile` is constrained to approved repository-relative `spec/tla/*.tla` paths before external formal tools run.
- Manual `alloyJar` / `tlaToolsJar` dispatch inputs are not exposed.
- Alloy shell command templates are not executed by `scripts/formal/verify-alloy.mjs`.
- Manual formal jobs require a write-like collaborator permission check before repository scripts run.

## Rollback

Revert this PR to restore the prior free-form formal dispatch and shell-template Alloy runner behavior.

### CODE REVIEW

File: .github/workflows/formal-verify.yml [hunk 1]
```diff
@@ -140,14 +168,21 @@ jobs:
...
+        run: node scripts/ci/assert-workflow-dispatch-actor.mjs
       - name: Setup Node + pnpm
         uses: ./.github/actions/setup-node-pnpm
         with:
           node-version: '20'
       - name: Run verify:tla stub
         env:
           TLA_ENGINE: ${{ github.event_name == 'workflow_dispatch' && inputs.engine || 'tlc' }}
-          TLA_TOOLS_JAR: ${{ github.event_name == 'workflow_dispatch' && inputs.tlaToolsJar || '' }}
+          TLA_TOOLS_JAR: ''
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/formal-verify.yml]: Addressed in head `055dd957`: the TLA job now installs Java, caches `.cache/tools/tla2tools.jar`, downloads it from the workflow-owned pinned `TLA_TOOLS_VERSION` URL when missing, and sets `TLA_TOOLS_JAR` internally. The dispatch jar override remains removed, so operators cannot inject a jar path, but the default `engine=tlc` path is runnable again.

File: scripts/ci/assert-workflow-dispatch-actor.mjs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+if (import.meta.url === `file://${process.argv[1]}`) {
+  try {
+    process.exit(main());
+  } catch (error) {
+    console.error(error?.message ?? String(error));
+    process.exit(1);
+  }
+}
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply scripts/ci/assert-workflow-dispatch-actor.mjs]: Addressed in head `055dd957`: the entrypoint guard now derives the argv script path through `path.resolve()` + `pathToFileURL()` before comparing it with `import.meta.url`. I also added unit coverage for direct CLI invocation and a smoke check now exits non-zero when the required workflow actor env is absent, confirming `main()` runs.
